### ISSUE TITLE
ktlint removes first U+FEFF seen anywhere in its input, even if it's in a literal

### ISSUE THREAD

**Issue by MattWindsor91** [pinterest/ktlint#3220]
https://github.com/pinterest/ktlint/pull/630 made ktlint temporarily remove byte order marks, and later put them back at the start of the file.  The way the removal in particular works seems to produce corner cases whereby ktlint can accidentally remove U+FEFF where it shouldn't, and not put them back.

When this U+FEFF is inside a char or string literal, this is unsound: the contents of the literal are changed by formatting.  Worse, if it's inside a char literal, the formatting process makes the literal empty and therefore _syntactically_ invalid.

While this problem relates to a character known to be problematic and to-be-avoided in UTF8 input, and can be worked around by always using escape sequences, it can be triggered for example (and this is how I encountered the bug!) by code generators that emit lightly-escaped literal UTF-8 data into string literals.

This behaviour seems to originate from [here](https://github.com/pinterest/ktlint/blob/76ecdfc9d18c49ff15fb9394b069493e8b5660c1/ktlint-rule-engine/src/main/kotlin/com/pinterest/ktlint/rule/engine/internal/RuleExecutionContext.kt#L262-L266), but as an outsider I suspect I may be missing some context as to why the problem of BOMs appearing in input code was solved this way.

## Expected Behavior

ktlint should only disturb U+FEFF at the start of an input script, as (to the best of my knowledge) this is the only place where it could possibly be a byte-order mark.  In no circumstances should it remove U+FEFF in char and string literals.

## Observed Behavior

ktlint can remove, and fail to replace, the first U+FEFF in the input script, even if it isn't at the start of the input.

## Steps to Reproduce

1. Create a KotlinScript file with both a code style violation and a `U+FEFF` character (not an escape sequence) inside something (such as a string or char literal) that shouldn't be changed by ktlint.  The example I have is:

```kotlin
println ('﻿')
```
(where the space before `(` is a style violation, and there is a `U+FEFF` character between the parentheses; IDEA shows this as `[ZWNBSP]`).

This produces the following hex dump:

```shell
$ xxd bug.kts
00000000: 7072 696e 746c 6e20 2827 efbb bf27 290a  println ('...').
```

Note the `efbb bf` in the hex dump, which is the UTF-8 encoding of `U+FEFF`.

In IDEA, this file is syntactically valid and executable, with the expected result.

2. Run ktlint in format mode on the file:

```shell
$ ktlint -F bug.kts
```

3. Observe (eg. in a hex editor or in IDEA) that the code has been formatted to

```kotlin
println('')
```

but the `U+FEFF` character has disappeared:

```shell
$ xxd bug.kts
00000000: 7072 696e 746c 6e28 2727 290a            println('').
```

This file is now syntactically invalid and can't be run from within IDEA:

```
bug.kts:1:9: error: empty character literal.
println('')
        ^
```

Note that if the file _starts_ with a BOM, ktlint works as expected (fixes the formatting but preserves all `U+FEFF`):

```shell
$  xxd bug.kts
00000000: efbb bf70 7269 6e74 6c6e 2028 27ef bbbf  ...println ('...
00000010: 2729 0a                                  ').
$ ktlint -F bug.kts
$ xxd bug.kts
00000000: efbb bf70 7269 6e74 6c6e 2827 efbb bf27  ...println('...'
00000010: 290a                                     ).
```

## Your Environment

```shell
$ ktlint --version
ktlint version 1.8.0
```

Observed in CLI usage with no `.editorconfig` on macOS Sequoia 15.7.3, system java is openjdk 11.0.24 2024-07-16 LTS.  Originally seen in a custom integration of a `KtLintRuleEngine` against ktlint 1.4.0.

**Comment by paul-dingemans** [comment]
Thank you for your elaborate issue report! Will fix this in the next release.
